File: sdk/javascript/agentlens.js
/**
 * AgentLens SDK for JavaScript/TypeScript (Node.js + Browser)
 * 
 * Universal HTTP-based SDK — works anywhere fetch() is available.
 * No dependencies required (uses native fetch in Node 18+ and browsers).
 * 
 * Usage:
 *   const lens = new AgentLens('http://localhost:8340', 'al_your_api_key');
 *   await lens.trackLLMCall({ model: 'gpt-4o', prompt: 'Hello', completion: 'Hi!', latency_ms: 120 });
 * 
 * For older Node.js (< 18), install node-fetch:
 *   npm install node-fetch
 */

class AgentLens {
    /**
     * @param {string} serverUrl - AgentLens server URL (e.g. http://localhost:8340)
     * @param {string} [apiKey] - API key (optional if auth not required)
     * @param {object} [options] - Additional options
     * @param {string} [options.projectId] - Project ID for multi-tenancy
     * @param {string} [options.agentName] - Default agent name
     * @param {number} [options.batchSize] - Events per batch (default: 50)
     * @param {number} [options.flushInterval] - Flush interval in ms (default: 5000)
     */
    constructor(serverUrl, apiKey = null, options = {}) {
        this.serverUrl = serverUrl.replace(/\/$/, '');
        this.apiKey = apiKey;
        this.projectId = options.projectId || 'default';
        this.agentName = options.agentName || 'default';
        this.batchSize = options.batchSize || 50;
        this.flushInterval = options.flushInterval || 5000;

        this._buffer = [];
        this._sessionId = null;
        this._timer = null;
        this._startAutoFlush();
    }

    // ─── Headers ────────────────────────────────────────────

    _headers() {
        const h = { 'Content-Type': 'application/json' };
        if (this.apiKey) h['Authorization'] = `Bearer ${this.apiKey}`;
        if (this.projectId) h['X-Project'] = this.projectId;
        return h;
    }

    // ─── Session Management ─────────────────────────────────

    startSession(agentName = null) {
        this._sessionId = `sess_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
        this._addEvent({
            event_type: 'session.start',
            session_id: this._sessionId,
            agent_name: agentName || this.agentName,
        });
        return this._sessionId;
    }

    endSession(success = true, meta = {}) {
        if (!this._sessionId) return;
        this._addEvent({
            event_type: 'session.end',
            session_id: this._sessionId,
            success,
            meta,
        });
        this._sessionId = null;
        this.flush(); // Flush immediately on session end
    }

    // ─── Event Tracking ─────────────────────────────────────

    trackLLMCall({ model, provider, prompt, completion, inputTokens, outputTokens, costUsd, latencyMs, ...extra }) {
        this._addEvent({
            event_type: 'llm.response',
            model,
            provider,
            prompt,
            completion,
            input_tokens: inputTokens,
            output_tokens: outputTokens,
            total_tokens: (inputTokens || 0) + (outputTokens || 0),
            cost_usd: costUsd,
            latency_ms: latencyMs,
            ...extra,
        });
    }

    trackToolCall({ toolName, toolArgs, toolResult, success = true, durationMs, ...extra }) {
        this._addEvent({
            event_type: success ? 'tool.result' : 'tool.error',
            tool_name: toolName,
            tool_args: toolArgs,
            tool_result: typeof toolResult === 'string' ? toolResult : JSON.stringify(toolResult),
            success,
            duration_ms: durationMs,
            ...extra,
        });
    }

    trackStep({ stepNumber, thought, decision, ...extra }) {
        this._addEvent({
            event_type: 'agent.step',
            step_number: stepNumber,
            thought,
            decision,
            ...extra,
        });
    }

    trackError({ errorType, errorMessage, stackTrace, ...extra }) {
        this._addEvent({
            event_type: 'error',
            error_type: errorType,
            error_message: errorMessage,
            stack_trace: stackTrace,
            success: false,
            ...extra,
        });
    }

    trackCustom(eventData) {
        this._addEvent({ event_type: 'custom', ...eventData });
    }

    // ─── Buffering & Flushing ───────────────────────────────

    _addEvent(event) {
        event.timestamp = event.timestamp || Date.now() / 1000;
        event.session_id = event.session_id || this._sessionId || '';
        event.agent_name = event.agent_name || this.agentName;
        event.event_id = `evt_${Date.now()}_${Math.random().toString(36).substr(2, 6)}`;
        this._buffer.push(event);

        if (this._buffer.length >= this.batchSize) {
            this.flush();
        }
    }

    async flush() {
        if (this._buffer.length === 0) return;

        const events = [...this._buffer];
        this._buffer = [];

        try {
            const resp = await fetch(`${this.serverUrl}/api/v1/events`, {
                method: 'POST',
                headers: this._headers(),
                body: JSON.stringify({ events }),
            });

            if (!resp.ok) {
                const err = await resp.text();
                console.error(`[AgentLens] Flush failed (${resp.status}): ${err}`);
                // Re-add events to buffer for retry
                this._buffer.unshift(...events);
            }
        } catch (e) {
            console.error(`[AgentLens] Flush error: ${e.message}`);
            this._buffer.unshift(...events);
        }
    }

    _startAutoFlush() {
        this._timer = setInterval(() => this.flush(), this.flushInterval);
    }

    async shutdown() {
        if (this._timer) {
            clearInterval(this._timer);
            this._timer = null;
        }
        await this.flush();
    }

    // ─── Query API ──────────────────────────────────────────

    async getSessions(options = {}) {
        const params = new URLSearchParams();
        if (options.agent) params.set('agent', options.agent);
        if (options.limit) params.set('limit', options.limit);
        if (options.offset) params.set('offset', options.offset);

        const resp = await fetch(`${this.serverUrl}/api/v1/sessions?${params}`, {
            headers: this._headers(),
        });
        return resp.json();
    }

    async getAnalytics(hours = 24) {
        const resp = await fetch(`${this.serverUrl}/api/v1/analytics?hours=${hours}`, {
            headers: this._headers(),
        });
        return resp.json();
    }

    async getHealth() {
        const resp = await fetch(`${this.serverUrl}/api/health`);
        return resp.json();
    }
}

// ─── OpenAI Wrapper (auto-instrument) ───────────────────────

/**
 * Wraps an OpenAI client to auto-track all LLM calls.
 * 
 * Usage:
 *   import OpenAI from 'openai';
 *   const openai = new OpenAI();
 *   const lens = new AgentLens('http://localhost:8340', 'al_key');
 *   const trackedOpenAI = lens.wrapOpenAI(openai);
 *   // Now all calls are auto-tracked
 */
AgentLens.prototype.wrapOpenAI = function(openaiClient) {
    const lens = this;
    const original = openaiClient.chat.completions.create.bind(openaiClient.chat.completions);

    openaiClient.chat.completions.create = async function(...args) {
        const start = Date.now();
        try {
            const result = await original(...args);
            const latency = Date.now() - start;
            const model = args[0]?.model || 'unknown';
            const prompt = JSON.stringify(args[0]?.messages || []);

            lens.trackLLMCall({
                model,
                provider: 'openai',
                prompt: prompt.substring(0, 2000),
                completion: result.choices?.[0]?.message?.content?.substring(0, 2000),
                inputTokens: result.usage?.prompt_tokens,
                outputTokens: result.usage?.completion_tokens,
                latencyMs: latency,
            });

            return result;
        } catch (error) {
            lens.trackError({
                errorType: 'openai_error',
                errorMessage: error.message,
                stackTrace: error.stack,
            });
            throw error;
        }
    };

    return openaiClient;
};

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { AgentLens };
}
if (typeof window !== 'undefined') {
    window.AgentLens = AgentLens;
}
export { AgentLens };

File: sdk/go/agentlens.go
// Package agentlens provides a Go SDK for AgentLens — AI Agent Observability.
//
// Usage:
//
//	lens := agentlens.New("http://localhost:8340", "al_your_api_key")
//	defer lens.Shutdown()
//
//	sess := lens.StartSession("my-go-agent")
//	lens.TrackLLMCall(agentlens.LLMEvent{
//	    Model: "gpt-4o", Prompt: "Hello", Completion: "Hi!", LatencyMs: 120,
//	})
//	lens.EndSession(sess, true, nil)
package agentlens

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"sync"
	"time"
)

// Client is the main AgentLens SDK client.
type Client struct {
	serverURL     string
	apiKey        string
	projectID     string
	agentName     string
	batchSize     int
	flushInterval time.Duration

	buffer    []map[string]interface{}
	mu        sync.Mutex
	sessionID string
	client    *http.Client
	done      chan struct{}
}

// Option configures the client.
type Option func(*Client)

func WithProjectID(id string) Option   { return func(c *Client) { c.projectID = id } }
func WithAgentName(name string) Option  { return func(c *Client) { c.agentName = name } }
func WithBatchSize(n int) Option        { return func(c *Client) { c.batchSize = n } }
func WithFlushInterval(d time.Duration) Option {
	return func(c *Client) { c.flushInterval = d }
}

// New creates a new AgentLens client.
func New(serverURL, apiKey string, opts ...Option) *Client {
	c := &Client{
		serverURL:     serverURL,
		apiKey:        apiKey,
		projectID:     "default",
		agentName:     "default",
		batchSize:     50,
		flushInterval: 5 * time.Second,
		buffer:        make([]map[string]interface{}, 0, 64),
		client:        &http.Client{Timeout: 10 * time.Second},
		done:          make(chan struct{}),
	}
	for _, o := range opts {
		o(c)
	}
	go c.autoFlush()
	return c
}

// StartSession begins a new monitoring session.
func (c *Client) StartSession(agentName string) string {
	if agentName == "" {
		agentName = c.agentName
	}
	c.mu.Lock()
	c.sessionID = fmt.Sprintf("sess_%d_%s", time.Now().UnixMilli(), randStr(8))
	sid := c.sessionID
	c.mu.Unlock()

	c.addEvent(map[string]interface{}{
		"event_type": "session.start",
		"session_id": sid,
		"agent_name": agentName,
	})
	return sid
}

// EndSession ends the current session.
func (c *Client) EndSession(sessionID string, success bool, meta map[string]interface{}) {
	if meta == nil {
		meta = map[string]interface{}{}
	}
	c.addEvent(map[string]interface{}{
		"event_type": "session.end",
		"session_id": sessionID,
		"success":    success,
		"meta":       meta,
	})
	c.mu.Lock()
	c.sessionID = ""
	c.mu.Unlock()
	c.Flush()
}

// LLMEvent represents an LLM call to track.
type LLMEvent struct {
	Model        string  `json:"model"`
	Provider     string  `json:"provider"`
	Prompt       string  `json:"prompt"`
	Completion   string  `json:"completion"`
	InputTokens  int     `json:"input_tokens"`
	OutputTokens int     `json:"output_tokens"`
	CostUSD      float64 `json:"cost_usd"`
	LatencyMs    float64 `json:"latency_ms"`
}

// TrackLLMCall records an LLM call event.
func (c *Client) TrackLLMCall(e LLMEvent) {
	c.addEvent(map[string]interface{}{
		"event_type":   "llm.response",
		"model":        e.Model,
		"provider":     e.Provider,
		"prompt":       e.Prompt,
		"completion":   e.Completion,
		"input_tokens": e.InputTokens,
		"output_tokens": e.OutputTokens,
		"total_tokens":  e.InputTokens + e.OutputTokens,
		"cost_usd":     e.CostUSD,
		"latency_ms":   e.LatencyMs,
	})
}

// ToolEvent represents a tool call to track.
type ToolEvent struct {
	ToolName   string                 `json:"tool_name"`
	ToolArgs   map[string]interface{} `json:"tool_args"`
	ToolResult string                 `json:"tool_result"`
	Success    bool                   `json:"success"`
	DurationMs float64               `json:"duration_ms"`
}

// TrackToolCall records a tool call event.
func (c *Client) TrackToolCall(e ToolEvent) {
	eventType := "tool.result"
	if !e.Success {
		eventType = "tool.error"
	}
	c.addEvent(map[string]interface{}{
		"event_type":  eventType,
		"tool_name":   e.ToolName,
		"tool_args":   e.ToolArgs,
		"tool_result": e.ToolResult,
		"success":     e.Success,
		"duration_ms": e.DurationMs,
	})
}

// TrackError records an error event.
func (c *Client) TrackError(errorType, errorMsg, stackTrace string) {
	c.addEvent(map[string]interface{}{
		"event_type":    "error",
		"error_type":    errorType,
		"error_message": errorMsg,
		"stack_trace":   stackTrace,
		"success":       false,
	})
}

// TrackStep records an agent reasoning step.
func (c *Client) TrackStep(stepNumber int, thought, decision string) {
	c.addEvent(map[string]interface{}{
		"event_type":  "agent.step",
		"step_number": stepNumber,
		"thought":     thought,
		"decision":    decision,
	})
}

func (c *Client) addEvent(event map[string]interface{}) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if _, ok := event["timestamp"]; !ok {
		event["timestamp"] = float64(time.Now().UnixMilli()) / 1000.0
	}
	if _, ok := event["session_id"]; !ok {
		event["session_id"] = c.sessionID
	}
	if _, ok := event["agent_name"]; !ok {
		event["agent_name"] = c.agentName
	}
	event["event_id"] = fmt.Sprintf("evt_%d_%s", time.Now().UnixMilli(), randStr(6))

	c.buffer = append(c.buffer, event)
	if len(c.buffer) >= c.batchSize {
		go c.Flush()
	}
}

// Flush sends all buffered events to the server.
func (c *Client) Flush() error {
	c.mu.Lock()
	if len(c.buffer) == 0 {
		c.mu.Unlock()
		return nil
	}
	events := c.buffer
	c.buffer = make([]map[string]interface{}, 0, 64)
	c.mu.Unlock()

	body := map[string]interface{}{"events": events}
	data, err := json.Marshal(body)
	if err != nil {
		// Re-add to buffer
		c.mu.Lock()
		c.buffer = append(events, c.buffer...)
		c.mu.Unlock()
		return fmt.Errorf("marshal error: %w", err)
	}

	req, err := http.NewRequest("POST", c.serverURL+"/api/v1/events", bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	if c.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+c.apiKey)
	}
	if c.projectID != "" {
		req.Header.Set("X-Project", c.projectID)
	}

	resp, err := c.client.Do(req)
	if err != nil {
		c.mu.Lock()
		c.buffer = append(events, c.buffer...)
		c.mu.Unlock()
		return fmt.Errorf("send error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return fmt.Errorf("server returned %d", resp.StatusCode)
	}
	return nil
}

// Shutdown flushes remaining events and stops the background goroutine.
func (c *Client) Shutdown() {
	close(c.done)
	c.Flush()
}

func (c *Client) autoFlush() {
	ticker := time.NewTicker(c.flushInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			c.Flush()
		case <-c.done:
			return
		}
	}
}

func randStr(n int) string {
	const letters = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, n)
	for i := range b {
		b[i] = letters[rand.Intn(len(letters))]
	}
	return string(b)
}
